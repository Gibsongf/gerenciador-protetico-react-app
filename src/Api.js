export async function apiLogin(loginData) {
    const url =
        "http://https://gerenciador-protetico.adaptable.app/users/login";
    // console.log(loginData);
    try {
        const response = await fetch(url, {
            method: "post",
            headers: {
                "Content-Type": "application/json",
            },
            body: JSON.stringify(loginData),
        });

        if (response.status === 200) {
            const data = await response.json();
            // console.log("Login successfully");
            localStorage.setItem("token", data.token);
            return true;
        } else {
            throw new Error();
        }
    } catch (error) {
        throw Error(error);
    }
}
export async function apiRegister(loginData) {
    const url =
        "http://https://gerenciador-protetico.adaptable.app/users/register";
    // console.log(loginData);
    try {
        const response = await fetch(url, {
            method: "post",
            headers: {
                "Content-Type": "application/json",
            },
            body: JSON.stringify(loginData),
        });

        if (response.status === 200) {
            const data = await response.json();
            // console.log("Login successfully");
            localStorage.setItem("token", data.token);
            return true;
        } else {
            throw new Error();
        }
    } catch (error) {
        throw Error(error);
    }
}
async function setupFetch(url, reqMethod = "get", body) {
    if (!localStorage["token"]) {
        await apiLogin();
        // localStorage.setItem("token", data.token);
    }

    const reqConfig = {
        method: reqMethod,
        headers: {
            "Content-Type": "application/json",
            Authorization: "Bearer " + localStorage["token"],
        },
    };
    if (body) {
        reqConfig.body = JSON.stringify(body);
    }
    try {
        const response = await fetch(url, reqConfig);
        if (response.status === 200 || response.status === 400) {
            const data = await response.json();
            return data;
        }

        if (response.status === 401) {
            localStorage.removeItem("token");
            return "Old token";
        }
    } catch (err) {
        throw Error(err);
    }
}
export async function APItodos(category) {
    const url = `http://https://gerenciador-protetico.adaptable.app/api/${category}/todos`;
    const data = await setupFetch(url, "get");
    // console.log(data);
    return data;
}
export async function APIDetails(category, id) {
    const url = `http://https://gerenciador-protetico.adaptable.app/api/${category}/${id}`;
    const data = await setupFetch(url, "get");
    return data;
}
export async function APIGetServiceBy(id, by) {
    const url = `http://https://gerenciador-protetico.adaptable.app/api/servico/todos/${by}/${id}`;
    const data = await setupFetch(url, "get");
    return data;
}

export async function APIPostNewData(formData) {
    const url = `http://https://gerenciador-protetico.adaptable.app/api/${formData["category"]}/novo`;
    // console.log(formData);
    const data = await setupFetch(url, "post", formData);
    // console.log(data);
    return data;
}
export async function APIPutData(formData, id) {
    const url = `http://https://gerenciador-protetico.adaptable.app/api/${formData["category"]}/${id}/edit`;
    // console.log(formData);
    const data = await setupFetch(url, "put", formData);
    // console.log(data);

    return data;
}
export async function ApiOneExcel(id) {
    const url = `http://https://gerenciador-protetico.adaptable.app/api/export/${id}`;
    const reqConfig = {
        method: "Get",
        headers: {
            Authorization: "Bearer " + localStorage["token"],
        },
    };

    try {
        const response = await fetch(url, reqConfig);
        if (response.status === 200) {
            const blob = await response.blob();
            // console.log(data);
            return blob;
        }
    } catch (err) {
        throw Error(err);
    }
}

export async function ApiMonthExcel(id, date) {
    const url = `http://https://gerenciador-protetico.adaptable.app/api/export/${id}/${date}`;
    const reqConfig = {
        method: "Get",
        headers: {
            Authorization: "Bearer " + localStorage["token"],
        },
    };

    try {
        const response = await fetch(url, reqConfig);
        if (response.status === 200) {
            const blob = await response.blob();
            // console.log(data);
            return blob;
        }
    } catch (err) {
        throw Error(err);
    }
}
