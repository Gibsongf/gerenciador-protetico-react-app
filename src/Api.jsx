export async function apiLogin(loginData) {
    const url = "http://localhost:3000/users/login";
    // console.log(loginData);
    try {
        const response = await fetch(url, {
            method: "post",
            headers: {
                "Content-Type": "application/json",
            },
            body: JSON.stringify(loginData),
        });

        if (response.status === 200) {
            const data = await response.json();
            console.log("Login successfully");
            localStorage.setItem("token", data.token);
            return true;
        } else {
            throw new Error();
        }
    } catch (error) {
        throw Error(error);
    }
}

async function setupFetch(url, reqMethod = "get", body) {
    if (!localStorage["token"]) {
        await apiLogin();
        // localStorage.setItem("token", data.token);
    }

    const reqConfig = {
        method: reqMethod,
        headers: {
            "Content-Type": "application/json",
            Authorization: "Bearer " + localStorage["token"],
        },
    };
    if (body) {
        reqConfig.body = JSON.stringify(body);
    }
    try {
        const response = await fetch(url, reqConfig);
        if (response.status === 200 || response.status === 400) {
            const data = await response.json();
            return data;
        }

        if (response.status === 401) {
            localStorage.removeItem("token");
            return "Old token";
        }
    } catch (err) {
        throw Error(err);
    }
}
export async function APItodos(category) {
    const url = `http://localhost:3000/api/${category}/todos`;
    const data = await setupFetch(url, "get");
    // console.log(data);
    return data;
}
export async function APIDetails(category, id) {
    const url = `http://localhost:3000/api/${category}/${id}`;
    const data = await setupFetch(url, "get");
    return data;
}
export async function APIGetServiceBy(id, by) {
    const url = `http://localhost:3000/api/servico/todos/${by}/${id}`;
    const data = await setupFetch(url, "get");
    return data;
}

export async function APIPostNewData(formData) {
    const url = `http://localhost:3000/api/${formData["category"]}/novo`;
    console.log(formData);
    const data = await setupFetch(url, "post", formData);
    // console.log(data);
    return data;
}
export async function APIPutData(formData, id) {
    const url = `http://localhost:3000/api/${formData["category"]}/${id}/edit`;
    // console.log(formData);
    const data = await setupFetch(url, "put", formData);
    // console.log(data);

    return data;
}
export async function ExcelLink(id) {
    const url = `http://localhost:3000/api/export/${id}`;
    const reqConfig = {
        method: "Get",
        headers: {
            Authorization: "Bearer " + localStorage["token"],
        },
    };

    try {
        const response = await fetch(url, reqConfig);
        if (response.status === 200) {
            const blob = await response.blob();
            // console.log(data);
            return blob;
        }
    } catch (err) {
        throw Error(err);
    }
}
