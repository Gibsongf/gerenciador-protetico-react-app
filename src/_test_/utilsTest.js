export const testService = {
    paciente: "Meire Macedo",
    dentista: "dentista-1",
    local: "local-1",
    produto: [
        {
            _id: "65dd5ff530d103b5e6f12059",
            nome: "Sem marca Granito Teclado",
            valor_normal: 132,
            valor_reduzido: 66,
            __v: 0,
        },
        {
            _id: "65dd5ff530d103b5e6f12053",
            nome: "Ergonômico Madeira Queijo",
            valor_normal: 200,
            valor_reduzido: 67,
            __v: 0,
        },
    ],
    statusEntrega: false,
    category: "servico",
    formType: "edit",
    dbId: "65dd5ff530d103b5e6f12067",
};
export const testDentist = {
    nome: "first dentist",
    sobrenome: "first sobrenome",
    telefone: "89224002",
    cpf: "10987654321",
    category: "dentista",
    formType: "edit",
    _id: "first-dentist-id",
    serviço: [],
};
export const testDentist2 = {
    nome: "second dentist",
    sobrenome: "second sobrenome",
    telefone: "40028922",
    cpf: "12345678910",
    category: "dentista",
    formType: "edit",
    _id: "second-dentist-id",
    serviço: [],
};
export const testLocal = [
    {
        _id: "local-0",
        nome: "Franco EIRELI",
        endereço: "333 Felícia Marginal",
        cep: "23147-272",
        telefone: "1174-0086",
        tabela: "Reduzido",
        dentistas: [testDentist],
        __v: 0,
    },
    {
        _id: "local-1",
        nome: "Moraes LTDA",
        endereço: "1047 Luiza Alameda",
        cep: "63006-953",
        telefone: "5541-7666",
        tabela: "Normal",
        dentistas: [testDentist2],
        __v: 0,
    },
];
