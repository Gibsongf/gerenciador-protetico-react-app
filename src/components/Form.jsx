import { useState } from "react";
import {
    CepInput,
    CpfInput,
    SelectInput,
    SimpleInput,
    TelefoneInput,
    TipoTabelaSelect,
} from "./Inputs";
import "../styles/Forms.css";
const UseForm = (formType, initialState) => {
    const [formData, setFormData] = useState(initialState);

    const handleChange = (e) => {
        // console.log(e.target);
        // console.log(formData);
        setFormData({ ...formData, [e.target.name]: e.target.value });
        // newContentValidator(e);
    };
    const handleSubmit = async (e) => {
        e.preventDefault();
        // const validData = newContentValidator(e.target.parentElement);

        // if (validData) {
        //     if (formType === "new") {
        //         // await newPost(formData);
        //         setFormData({ text: "", title: "", published: false });
        //     }
        //     if (formType === "edit") {
        //         setWasUpdated((e) => !e);
        //     }
        // }
    };
    return { formData, handleChange, handleSubmit };
};

export function FormLocal() {
    const nomeTag = { id: "nome", txt: "Nome do Local ou a Franquia:" };
    const initialState = {
        nome: "",
        endereço: "",
        cep: "",
        telefone: "",
        tabela: "",
    };
    const { formData, handleChange, handleSubmit } = UseForm(
        "new",
        initialState
    );

    return (
        <form action="">
            <legend>
                <h3>Registrar Novo Local</h3>
            </legend>
            <SimpleInput
                id={nomeTag.id}
                labelTxt={nomeTag.txt}
                value={formData.nome}
                onChange={handleChange}
            />
            <SimpleInput
                id={"endereço"}
                labelTxt={"Endereço:"}
                value={formData.endereço}
                onChange={handleChange}
            />
            <CepInput value={formData.cep} onChange={handleChange} />
            <TelefoneInput value={formData.telefone} onChange={handleChange} />
            <TipoTabelaSelect value={formData.tabela} onChange={handleChange} />
            <button type="submit">Registrar</button>
        </form>
    );
}
export function FormDentist() {
    const initialState = {
        nome: "",
        sobrenome: "",
        local: "",
        telefone: "",
        cpf: "",
    };

    const { formData, handleChange, handleSubmit } = UseForm(
        "new",
        initialState
    );

    return (
        <form action="">
            <legend>
                <h3>Registrar Novo Dentista</h3>
            </legend>
            <SimpleInput
                id={"nome"}
                labelTxt={"Nome:"}
                value={formData.nome}
                onChange={handleChange}
            />
            <SimpleInput
                id={"sobrenome"}
                labelTxt={"Sobrenome:"}
                value={formData.sobrenome}
                onChange={handleChange}
            />
            <TelefoneInput value={formData.telefone} onChange={handleChange} />
            <CpfInput value={formData.cpf} onChange={handleChange} />

            <SelectInput
                initialValue={formData.local}
                onChange={handleChange}
                category={"local"}
                labelTxt={"Local de Trabalho:"}
            />
            <button type="submit">Registrar</button>
        </form>
    );
}

export function FormProduct() {
    const initialState = {
        nome: "",
        valor_normal: "",
        valor_reduzido: "",
    };

    const { formData, handleChange, handleSubmit } = UseForm(
        "new",
        initialState
    );

    return (
        <form action="">
            <legend>
                <h3>Registrar Novo Produto</h3>
            </legend>
            <SimpleInput
                id={"nome"}
                labelTxt={"Nome:"}
                value={formData.nome}
                onChange={handleChange}
            />
            <SimpleInput
                id={"valor_normal"}
                labelTxt={"Valor Normal:"}
                value={formData.valor_normal}
                onChange={handleChange}
                type={"number"}
            />
            <SimpleInput
                id={"valor_reduzido"}
                labelTxt={"Valor Reduzido:"}
                value={formData.valor_reduzido}
                onChange={handleChange}
                type={"number"}
            />

            <button type="submit">Registrar</button>
        </form>
    );
}

export function FormService() {
    const initialState = {
        dentista: "",
        produto: "",
        paciente_nome: "",
    };

    const { formData, handleChange, handleSubmit } = UseForm(
        "new",
        initialState
    );

    return (
        <form action="">
            <legend>
                <h3>Registrar Novo Serviço</h3>
            </legend>
            <SimpleInput
                id={"paciente_nome"}
                labelTxt={"Nome do Paciente:"}
                value={formData.paciente_nome}
                onChange={handleChange}
            />
            {/* select Local then from this local we 
                show the dentist that work there as
                option for the next select input  
            */}
            <SelectInput
                initialValue={formData.dentista}
                onChange={handleChange}
                category={"local"}
                labelTxt={"Local:"}
            />

            <button type="submit">Registrar</button>
        </form>
    );
}
