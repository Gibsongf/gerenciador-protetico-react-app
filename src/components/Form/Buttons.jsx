import { useContext } from "react";
import PropTypes from "prop-types";
import { EditContext } from "../../pages/Details";

ButtonRegister.propTypes = {
    handleSubmit: PropTypes.func,
};
ButtonEdit.propTypes = {
    handleSubmit: PropTypes.func,
};
export function ButtonRegister({ handleSubmit }) {
    return (
        <button onClick={handleSubmit} type="submit">
            Registrar
        </button>
    );
}
export function ButtonEdit({ handleSubmit }) {
    const { setEdit } = useContext(EditContext);
    return (
        <>
            <button onClick={handleSubmit} type="submit">
                Confirm
            </button>

            <button type="button" onClick={() => setEdit((e) => !e)}>
                Cancel
            </button>
        </>
    );
}
