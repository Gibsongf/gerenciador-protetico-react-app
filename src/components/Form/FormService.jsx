import { useContext, useRef, useState } from "react";
import { v4 as uuidv4 } from "uuid";

import {
    FilterDentistsByLocation,
    SearchProducts,
    SelectInput,
    SimpleInput,
} from "../Inputs";

import "../../styles/Forms.css";
import { useTodosApi } from "../ApiHooks";
import { AppContext } from "../../App";
import { useForm } from "./useForm";

export function FormService() {
    const { errorMsg } = useContext(AppContext);
    const [produtoKeys, setProdutoKeys] = useState(["produto"]);
    const ref = useRef();

    const initialState = {
        dentista: "",
        local: "",
        paciente_nome: "",
        category: "servico",
    };
    const products = useTodosApi("produto");
    const [productInput, setProductInput] = useState([]);
    // //console.log(products);
    const { formData, handleChange, handleSubmit } = useForm(
        "new",
        initialState,
        ref.current,
        produtoKeys
    );

    const AdditionalProduct = () => {
        const key = uuidv4();
        setProductInput((previous) => {
            const name = "produto" + String(previous.length + 1);
            setProdutoKeys((prev) => [...prev, name]);
            return [
                ...previous,
                <SearchProducts
                    products={products}
                    onChange={handleChange}
                    key={key}
                    name={name}
                />,
            ];
        });
    };
    // //console.log(errorMsg);
    return (
        <form action="" ref={ref}>
            <legend>
                <h3>Registrar Novo Servi√ßo</h3>
            </legend>
            <SimpleInput
                id={"paciente_nome"}
                labelTxt={"Nome do Paciente:"}
                value={formData.paciente_nome}
                onChange={handleChange}
                msg={!errorMsg ? "" : errorMsg.paciente_nome}
            />

            <SelectInput
                initialValue={formData.local}
                onChange={handleChange}
                category={"local"}
                labelTxt={"Local:"}
                msg={!errorMsg ? "" : errorMsg.local}
            />

            <FilterDentistsByLocation
                initialValue={formData.dentista}
                onChange={handleChange}
                dbId={formData.local}
                msg={!errorMsg ? "" : errorMsg.dentista}
            />
            <SearchProducts
                products={products}
                name="produto"
                onChange={handleChange}
            />
            {productInput.map((i) => i)}

            <button onClick={AdditionalProduct} type="button">
                Mais Produto
            </button>
            <button onClick={handleSubmit} type="submit">
                Registrar
            </button>
        </form>
    );
}
