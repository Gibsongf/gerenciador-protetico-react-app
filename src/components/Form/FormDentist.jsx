import { useContext, useRef, useState } from "react";

import { CpfInput, SelectInput, SimpleInput, TelefoneInput } from "../Inputs";
import "../../styles/Forms.css";
import PropTypes from "prop-types";

import { useForm } from "./useForm";
import { AppContext } from "../../App";
import { ButtonEdit, ButtonRegister } from "./Buttons";
import { NewFormContext } from "../NewFormButton";
import { EditContext } from "../GenerateDetails";

export function FormDentist({ initialState, closeBtn }) {
    const ref = useRef();
    const { errorMsg } = useContext(AppContext);
    const { setClose, setTableUpdate } = useContext(NewFormContext);
    const { setEdit, setUpdate } = useContext(EditContext);
    let legendTxt = "Editar Detalhes do Dentista";
    if (!initialState) {
        legendTxt = "Registrar Novo Dentista";
        initialState = {
            nome: "",
            sobrenome: "",
            local: "",
            telefone: "",
            cpf: "",
            category: "dentista",
            formType: "new",
        };
    }

    const { formData, handleChange, handleSubmit } = useForm(
        initialState,
        ref.current
    );
    const handleSubmitResponse = async (e) => {
        e.preventDefault();

        const success = await handleSubmit(e);
        if (success) {
            if (initialState.formType === "new") {
                setClose((e) => !e);
                setTableUpdate((e) => !e);
            }
            if (initialState.formType === "edit") {
                setEdit((e) => !e);
                setUpdate((e) => !e);
            }
        }
    };
    return (
        <div className="form-container" id="pop-up">
            <form action="" ref={ref} id="pop-up-content">
                {closeBtn}
                <legend>
                    <h3>{legendTxt}</h3>
                </legend>
                <SimpleInput
                    id={"nome"}
                    labelTxt={"Nome"}
                    value={formData.nome}
                    onChange={handleChange}
                    msg={!errorMsg ? "" : errorMsg.nome}
                />
                <SimpleInput
                    id={"sobrenome"}
                    labelTxt={"Sobrenome"}
                    value={formData.sobrenome}
                    onChange={handleChange}
                />
                <TelefoneInput
                    value={formData.telefone}
                    onChange={handleChange}
                    msg={!errorMsg ? "" : errorMsg.telefone}
                />
                <CpfInput
                    value={formData.cpf}
                    onChange={handleChange}
                    msg={!errorMsg ? "" : errorMsg.cpf}
                />

                <SelectInput
                    initialValue={formData.local}
                    onChange={handleChange}
                    category={"local"}
                    labelTxt={"Local de Trabalho"}
                    msg={!errorMsg ? "" : errorMsg.local}
                />
                {initialState.formType === "edit" && (
                    <ButtonEdit handleSubmit={handleSubmitResponse} />
                )}
                {initialState.formType === "new" && (
                    <ButtonRegister handleSubmit={handleSubmitResponse} />
                )}
            </form>
        </div>
    );
}

FormDentist.propTypes = {
    initialState: PropTypes.object,
    closeBtn: PropTypes.any,
};
