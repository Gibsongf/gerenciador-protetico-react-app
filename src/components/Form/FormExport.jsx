import { createContext, useContext, useRef } from "react";

import {
    EntregaStatus,
    FilterDentistsByLocation,
    SearchProducts,
    SelectInput,
    SimpleInput,
} from "../Inputs";
import PropTypes from "prop-types";
import "../../styles/Forms.css";
import { useTodosApi } from "../ApiHooks";
import { AppContext } from "../../App";
import { useForm } from "./useForm";
import { ButtonConfirm, ButtonEdit, ButtonRegister } from "../Buttons";
import { NewFormContext } from "../NewFormButton";
import { EditContext } from "../GenerateDetails";
import { PopUpEditContext } from "../../pages/Todos";

FormExport.propTypes = {
    initialState: PropTypes.any,
    closeBtn: PropTypes.any,
};
export const RefContext = createContext({
    checkBoxRef: "",
});
export function FormExport({ initialState, closeBtn }) {
    const { setExport } = useContext(PopUpEditContext);

    // const products = useTodosApi("produto");
    // const { formData, setFormData, handleChange, handleSubmit } = useForm(
    //     initialState,
    //     ref.current
    // );

    // use the ref context to filter the selected checkbox and
    //format to array and save to the form data

    // console.log(initialState, formData);
    const beforeSendSubmit = async (e) => {
        e.preventDefault();
        //before submit we format the product

        // const success = await handleSubmit(e);
        // if (success) {
        //     if (initialState.formType === "new") {
        //         setClose((e) => !e);
        //         setTableUpdate((e) => !e);
        //     }
        //     if (initialState.formType === "edit") {
        //         setShowForm(() => false);
        //         setUpdate((e) => !e);
        //     }
        // }
    };

    return (
        <div className="form-container" id="pop-up">
            <form action="" id="pop-up-content">
                {closeBtn}
                <legend>
                    <h3>Configurar Exportação</h3>
                </legend>
                <label htmlFor="month">
                    / Mês:{" "}
                    <input
                        style={{ width: "160px" }}
                        // ref={ref}
                        // onChange={onChangeMonth}
                        lang="pt-BR"
                        type="month"
                        name="month"
                        id="month"
                    />
                </label>
                {/* <SimpleInput
                    id={"paciente"}
                    labelTxt={"Nome do Paciente:"}
                    value={formData.paciente}
                    onChange={handleChange}
                    msg={!errorMsg ? "" : errorMsg.paciente}
                /> */}

                {/* <SelectInput
                    initialValue={formData.local}
                    onChange={handleChange}
                    category={"local"}
                    labelTxt={"Local:"}
                    msg={!errorMsg ? "" : errorMsg.local}
                /> */}

                {/* <FilterDentistsByLocation
                    initialValue={formData.dentista}
                    onChange={handleChange}
                    dbId={formData.local}
                    msg={!errorMsg ? "" : errorMsg.dentista}
                /> */}

                <EntregaStatus
                    value={formData.statusEntrega}
                    onChange={handleChange}
                />
            </form>
        </div>
    );
}
