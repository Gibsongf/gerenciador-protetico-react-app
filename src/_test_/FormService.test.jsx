/* eslint-disable no-undef */
import { render, screen } from "@testing-library/react";
import { expect, vi } from "vitest";
import userEvent from "@testing-library/user-event";
import { APIPostNewData, APIPutData } from "../Api";
import { FormService } from "../components/Form/FormService";
import { testLocal, testService } from "./utilsTest";

//need to pass api mock to another file from all form tests
const mockApiData = {
    produto: [
        {
            nome: "produto-1",
            _id: "id-produto-1",
        },
        {
            nome: "produto-2",
            _id: "id-produto-2",
        },
    ],
    local: testLocal,
};
vi.mock("../components/ApiHooks", () => {
    return {
        useTodosApi: vi.fn((e) => {
            if (e) {
                return mockApiData[e];
            }
            return testLocal;
        }),
        useDetailsApi: vi.fn((e, id) => {
            if (e) {
                if (e === "local" && id) {
                    const indx = Number(id.split("-")[1]);
                    return { data: mockApiData[e][indx] };
                }
                return mockApiData[e];
            }
        }),
    };
});
vi.mock("../Api", () => {
    return {
        APIPostNewData: vi.fn(() => {
            return {
                errors: "",
            };
        }),
        APIPutData: vi.fn(() => {
            return {
                errors: "",
            };
        }),
        APIGetServiceBy: vi.fn(() => {
            return;
        }),
    };
});

//need to get id of some local and mock useForm return a local?
describe("Form Dentist component", () => {
    //all useful elements of the form
    const expectFormElements = (el) => {
        expect(el.paciente.value).toBe(testService.paciente);
        expect(el.local.value).toBe(testService.local);
        expect(el.dentista.value).toBe(testService.dentista);
        expect(el.checkBoxProduct.value).toBe(mockApiData.produto[0]._id);
    };
    const getEl = () => {
        const paciente = screen.getByLabelText("Nome do Paciente:");
        const local = screen.getByLabelText("Local:");
        const dentista = screen.getByLabelText("Dentistas:");
        const produto = screen.getByLabelText("Produto:");
        const btnConfirmProduct = screen.getByRole("button", {
            name: "Selecionar Produto",
        });
        const submit = screen.getAllByRole("button")[2];
        return {
            paciente,
            local,
            dentista,
            produto,
            btnConfirmProduct,
            submit,
        };
    };

    const inputText = async (el, text, user) => {
        await user.dblClick(el);
        await user.keyboard(text);
    };
    it("renders correctly", () => {
        const { container } = render(<FormService />);
        const form = screen.getByRole("service-form");
        expect(container).toMatchSnapshot();
        expect(form).toBeInTheDocument();
    });
    it("register all input, submit form & call API", async () => {
        const user = userEvent.setup();
        render(<FormService />);

        const { paciente, local, dentista, btnConfirmProduct, submit } =
            getEl();
        const header = screen.getByRole("heading", {
            name: "Registrar Novo Serviço",
        });
        const optionProduct = screen.getAllByRole("combobox")[2];

        await inputText(paciente, testService.paciente, user);
        await user.selectOptions(local, ["local-0"]);
        await user.selectOptions(dentista, ["first-dentist-id"]);
        await user.selectOptions(optionProduct, ["id-produto-1"]);

        //after select product click to add it to that service
        await user.click(btnConfirmProduct);
        await user.click(submit);

        //show just the dentist of the selected Local
        const localDentist = await screen.findByText("first dentist");

        //render a checkbox after click button to add product to the service
        const checkBoxProduct = screen.getByRole("checkbox", {
            name: mockApiData.produto[0].nome,
        });

        //header without initialState
        expect(header.textContent).toBe("Registrar Novo Serviço");

        //inputs has values correctly
        expectFormElements({ paciente, local, dentista, checkBoxProduct });

        //input with value of the dentist that work at selected Local is render
        expect(localDentist).toBeInTheDocument();

        //call the right API to put new data
        expect(APIPostNewData).toBeCalled();
    });
    it("Form edit mode, save data at right inputs", async () => {
        const user = userEvent.setup();

        render(<FormService initialState={testService} />);
        const header = screen.getByRole("heading", {
            name: "Editar Detalhes do Serviço",
        });
        const { paciente, local, dentista, submit } = getEl();

        await user.click(submit);

        //show just the dentist of the selected Local
        const localDentist = await screen.findByText("first dentist");
        //render a checkbox after click button to add product to the service
        const checkBoxProduct = screen.getByRole("checkbox", {
            name: mockApiData.produto[0].nome,
        });
        //header without initialState
        expect(header.textContent).toBe("Editar Detalhes do Serviço");

        //inputs has values correctly
        expectFormElements({ paciente, local, dentista, checkBoxProduct });

        //input with value of the dentist that work at selected Local is render
        expect(localDentist).toBeInTheDocument();

        //call the right API to update data
        expect(APIPutData).toBeCalled();
    });
    it("submit empty form, won't call API", async () => {
        const user = userEvent.setup();
        render(<FormService />);
        const { submit } = getEl();
        await user.click(submit);

        expect(APIPostNewData).not.toBeCalled();
    });
});
