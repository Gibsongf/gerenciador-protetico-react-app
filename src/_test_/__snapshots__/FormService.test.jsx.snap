// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Form Dentist component > renders correctly 1`] = `
<div>
  <div
    class="form-container"
    id="pop-up"
  >
    <form
      action=""
      id="pop-up-content"
      role="service-form"
    >
      <span
        class="form-close-btn"
      >
        <button
          class="close-btn"
          type="button"
        >
          X
        </button>
      </span>
      <legend>
        <h3>
          Registrar Novo Serviço
        </h3>
      </legend>
      <div
        class="paciente"
      >
        <label
          for="paciente"
        >
          Nome do Paciente
        </label>
        <div
          class="error-message"
        />
        <input
          id="paciente"
          name="paciente"
          required=""
          type="text"
          value=""
        />
      </div>
      <div
        class="select-options"
      >
        <label
          for="local"
        >
          Local
        </label>
        <div
          class="error-message"
        />
        <select
          id="local"
          name="local"
        >
          <option />
          <option
            value="local-0"
          >
            Franco EIRELI
          </option>
          <option
            value="local-1"
          >
            Moraes LTDA
          </option>
        </select>
      </div>
      <div
        class="select-options"
      >
        <label
          for="dentista"
        >
          Dentistas
        </label>
        <div
          class="error-message"
        />
        <select
          id="dentista"
          name="dentista"
          required=""
        >
          <option />
        </select>
      </div>
      <div
        class="search-product"
      >
        <label
          for="search-product"
        >
          Produto
        </label>
        <div
          class="error-message"
        />
        <input
          id="search-product"
          name="search-product"
          placeholder="Search..."
          type="text"
          value=""
        />
        <select
          id="produto"
          name="produto"
        >
          <option
            value=""
          />
          <option
            value="id-produto-1"
          >
            produto-1
          </option>
          <option
            value="id-produto-2"
          >
            produto-2
          </option>
        </select>
        <button
          class="select-product"
          type="button"
        >
          Selecionar Produto
        </button>
        <div
          for="selected-products"
        >
          Selecionado:
        </div>
      </div>
      <div
        class="statusEntrega"
      >
        <label
          for="statusEntrega"
        >
          Finalizado:
        </label>
        <select
          id="statusEntrega"
          name="statusEntrega"
        >
          <option
            value="false"
          >
            Não
          </option>
          <option
            value="true"
          >
            Sim
          </option>
        </select>
      </div>
      <button
        class="register-form-btn"
        type="submit"
      >
        Registrar
      </button>
    </form>
  </div>
</div>
`;
